### PR TITLE
Apple mobileconfig enhancements

### PR DESCRIPTION
**Pull request by feldsam**
Signed-off-by: Kristián Feldsam <feldsam@gmail.com>

### CODE REVIEW

**Comment by mkuron** [comment]
The changes look sensible. However, I don't think mounting the SSL keys directory into the PHP-FPM container is a good idea in terms of security (and it doesn't matter whether the private key is password-protected or not). The PHP interpreter is notorious for occasionally having security bugs that compromise files accessible to the web server, or there could be a path traversal issue lurking somewhere in Mailcow's PHP scripts, or something like that. **We should not take any chances in potentially exposing the web and mail server's SSL key.**

To make this secure, you should create a separate subdomain (e.g. `mobileconfig.$MAILCOW_HOSTNAME`), request a _separate_ certificate for that, store it into a separate directory (e.g. ./data/assets/ssl-signing/) and only expose that to the PHP container. You should then modify the DNS diagnostics page so that it also checks that `mobileconfig.$MAILCOW_HOSTNAME` is a CNAME for `$MAILCOW_HOSTNAME`.

**Comment by feldsam** [comment]
Hi, thanks, can you participate too and help with that subdomain solution?

**Comment by feldsam** [comment]
Hello, I see demand for this PR to be merged. @mkuron would you please help me with that separate cert? or I split this PR to two (general updates and signing)

**Comment by mkuron** [comment]
@feldsam, I‘m sorry I didn’t have time for this yet. Splitting into two pull requests would be helpful so we can get the UUID changes etc. into the next release.

**Comment by andryyy** [comment]
The signing should additionally be done using PHPs OpenSSL functions.

A random UUID sounds good.

**Comment by feldsam** [comment]
Hi, I removed signing part from this PR, so ready to merge

**Comment by feldsam** [comment]
@andryyy what you think by php openssl functions? We still need some key, or not?

**Comment by andryyy** [comment]
Looks good, but don't just add a link to the mobileconfig php... I don't want people to complain about why their phone is now setup with IMAP instead of EAS. If you want this link to be available, set it up for both EAS and IMAP, please.
Also add a note that it is only meant for iOS devices.

**Comment by andryyy** [comment]
Regarding OpenSSL functions: I won't ever merge proc_open or shell_exec stuff. I'd like to disable these functions alltogether sometime soon.

**Comment by feldsam** [comment]
I see, ok, where I find EAS link? I don't know about EAS Apple autoconfig file.

proc_open - ok, I understand.

**Comment by andryyy** [comment]
"com.apple.eas.account as the PayloadType" => https://developer.apple.com/library/content/featuredarticles/iPhoneConfigurationProfileRef/Introduction/Introduction.html

How about adding a _GET parameter like ?imap and ?eas

**Comment by feldsam** [comment]
Ok, I do till monday resp. durring weekend

**Comment by andryyy** [comment]
I can help this weekend. Or  try to. Time... :-(

**Comment by mkuron** [comment]
I‘d just remove the link. We already link to the mobileconfig from the documentation. We don‘t have direct links to the setup of other clients in the web UI either.

For EAS, you don‘t really need a mobileconfig as it is simple enough to set up manually (through autodiscover).

**Comment by feldsam** [comment]
Ok, I removed it

**Comment by extremeshok** [comment]
What still needs to be done to get this merged ?
Once its done, we can work on adding the signing part using a separate signing domain (mobileconfig.domain)

File: data/web/inc/functions.inc.php [hunk 1]
```diff
@@ -1183,4 +1183,19 @@ function get_logs($container, $lines = false) {
   }
   return false;
 }
-?>
+function getGUID() {
+  if (function_exists('com_create_guid')) {
+    return com_create_guid();
+  }
+
+  mt_srand((double)microtime()*10000);//optional for php 4.2.0 and up.
```
Comment (andryyy) [inline data/web/inc/functions.inc.php:1191]: What is that good for?

Comment (feldsam) [reply data/web/inc/functions.inc.php:1191]: Seeds the Mersenne Twister Random Number Generator

Comment (feldsam) [reply data/web/inc/functions.inc.php]: but for just this project we can use just com_create_guid()

File: data/web/inc/functions.inc.php [hunk 2]
```diff
@@ -1183,19 +1183,8 @@ function get_logs($container, $lines = false) {
...
-  mt_srand((double)microtime()*10000);//optional for php 4.2.0 and up.
-  $charid = strtoupper(md5(uniqid(rand(), true)));
-  $hyphen = chr(45);// "-"
-
-  return substr($charid, 0, 8).$hyphen
-        .substr($charid, 8, 4).$hyphen
-        .substr($charid,12, 4).$hyphen
-        .substr($charid,16, 4).$hyphen
-        .substr($charid,20,12);
+function getGUID($id) {
```
Comment (andryyy) [inline data/web/inc/functions.inc.php]: To prevent adding multiple configs for the same identity.

Comment (feldsam) [reply data/web/inc/functions.inc.php]: ok

Comment (feldsam) [reply data/web/inc/functions.inc.php]: but it should not be GUID named

Comment (feldsam) [reply data/web/inc/functions.inc.php]: I think, that it is unnecessary change. According to docs `A globally unique identifier for the payload. The actual content is unimportant, but it must be globally unique. In macOS, you can use uuidgen to generate reasonable UUIDs.`

Comment (andryyy) [reply data/web/inc/functions.inc.php]: Using the same UID overrides profiles, doesn't it?
edit: Or it prevents them from being added.

Comment (feldsam) [reply data/web/inc/functions.inc.php]: you have duplicates in one profile now

Comment (feldsam) [reply data/web/inc/functions.inc.php]: I personaly not tested replacing profiles, but if yes, them should be only on line 160 and not on all UUIDs

Comment (feldsam) [reply data/web/inc/functions.inc.php]: I know, that if profile name is same, it not allow to install it.

Comment (feldsam) [reply data/web/inc/functions.inc.php]: I tested it on iPad, and my version with GUID works fine

Comment (feldsam) [reply data/web/inc/functions.inc.php]: I tested it now, it it don't allow to install same profile twice, no matter if guid si same or not. It just checks `PayloadIdentifier`

Comment (feldsam) [reply data/web/inc/functions.inc.php]: So I propose to revert that changes and keep for ex. just that `com_create_guid` function

Comment (andryyy) [reply data/web/inc/functions.inc.php]: Please respond in a single comment.....

Comment (andryyy) [reply data/web/inc/functions.inc.php]: Okay, I see. So a random GUID is not a problem, but the PayloadIdentifier needs to be different for each profile. I have seen an issue which erroneously made me assume it was the GUID, sorry.

Can you revert my changes in your pr?

Just use `<?= xy; ?>` and 2 space indents.

**Comment by feldsam** [comment]
I reverted your changes, so it is ok now, or do you want to change something? I can do that now

edit: I go to try use just com_create_guid
edit: Tested and found, that we don't have com_create_guid available in php, so I think that this can be final version. I tested various scenarios on my iPad and iPhone before contributing

edit: just one thing, I thinking about is ability to configure just IMAP, because not all customers wants to use calendar and contacts, especially when have multiple accounts

**Comment by andryyy** [comment]
Hm, if we do this, another version with EAS would be super nice! :-)

**Comment by feldsam** [comment]
no problem, but @mkuron have comment on it :) sooo I suppose to merge this PR and make new one for EAS and just IMAP config. To move things on.

**Comment by andryyy** [comment]
Well, he is not wrong, configuring EAS is very easy with autodiscover. Could be added as optional feature though. Can you do this? Like bla.php?eas etc.?

Please do me a favor and change `<? echo ` to `<?= bla` etc. and use 2 space indents :-) I will be afk for a few hours now, we can merge it then.

**Comment by feldsam** [comment]
Ok, I do

**Comment by feldsam** [comment]
@andryyy reminding myself

**Comment by andryyy** [comment]
@mkuron You think that's fine?

**Comment by mkuron** [comment]
@andryyy, looks good to me.

**Comment by vain90** [comment]
Hi,
this PR was only merged into the dev branch.
Are there any plans when this PR will be merged into the master branch?

**Comment by andryyy** [comment]
Just forgot about it. :-(

**Comment by strarsis** [comment]
So there is a PR for https://github.com/mailcow/mailcow-dockerized/issues/3681 ? Awesome, can this be merged into main branch?

**Pull request merged**